cosmwasm/enclaves/execute/src/registration/attestation.rs

```diff
@@ -289,6 +292,230 @@ pub fn get_mr_enclave() -> [u8; 32] {
     rsgx_self_report().body.mr_enclave.m
 }
 
+
+#[cfg(feature = "SGX_MODE_HW")]
+pub fn verify_quote_ecdsa(
+    vec_quote : &Vec<u8>,
+    vec_coll : &Vec<u8>,
+    time_s: i64,
+) -> Result<(sgx_report_body_t, sgx_ql_qv_result_t), sgx_status_t>
+{
+    let mut qe_report: sgx_ql_qe_report_info_t = sgx_ql_qe_report_info_t::default();
+    let mut p_supp: [u8; 5000] = [0; 5000];
+    let mut n_supp: u32 = 0;
+    let mut exp_time_s: i64 = 0;
+    let mut exp_status: u32 = 0;
+    let mut qv_result: sgx_ql_qv_result_t = sgx_ql_qv_result_t::default();
+    let mut rt: sgx_status_t = sgx_status_t::default();
+
+    let mut ti: sgx_target_info_t = sgx_target_info_t::default();
+    unsafe {
+        sgx_self_target(&mut ti)
+    };
+
+    let res = unsafe {
+        ocall_verify_quote_ecdsa(
+            &mut rt as *mut sgx_status_t,
+            vec_quote.as_ptr(),
+            vec_quote.len() as u32,
+            vec_coll.as_ptr(),
+            vec_coll.len() as u32,
+            &ti,
+            time_s,
+            &mut qe_report,
+            p_supp.as_mut_ptr(),
+            p_supp.len() as u32,
+            &mut n_supp,
+            &mut exp_time_s,
+            &mut exp_status,
+            &mut qv_result)
+    };
+
+    if res != sgx_status_t::SGX_SUCCESS {
+        return Err(res);
+    }
+    if rt != sgx_status_t::SGX_SUCCESS {
+        return Err(rt);
+    }
+
+    match qv_result {
+        sgx_ql_qv_result_t::SGX_QL_QV_RESULT_OK => {},
+        sgx_ql_qv_result_t::SGX_QL_QV_RESULT_SW_HARDENING_NEEDED => {},
+        _ => {
+            trace!("Quote verification result: {}", qv_result);
+            return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+        }
+    };
+
+    // verify the qve report
+    if time_s != 0 {
+        exp_time_s = time_s; // insist on our time, if supplied
+    }
+
+    let qve_isvsvn_threshold: sgx_isv_svn_t = 3;
+    let dcap_ret : sgx_quote3_error_t  = unsafe { sgx_tvl_verify_qve_report_and_identity(
+        vec_quote.as_ptr(),
+        vec_quote.len() as u32,
+        &qe_report,
+        exp_time_s,
+        exp_status,
+        qv_result,
+        p_supp.as_ptr(),
+        n_supp,
+        qve_isvsvn_threshold) };
+
```

Comment:
Does this verification happen inside an enclave? Where do we cryptographically verify the attestation? i.e. how do we know someone isn't just faking a response

Reply:
In DCAP this is kinda complicated.

The verification consist of bringing all the relevant info (the original enclave quote, the so-called collateral, and the current time), and then verifying all this at-once.
The actual verification is performed by yet another enclave, provided by intel, called Quote Verificaiotn Enclave, QvE.

Here specifically the code that runs within our enclave. First it calls ocall_verify_quote_ecdsa() to trigger our code outside the enclave to check the attestation. It would load a QvE, provide it all the given information, the QvE would check all this and return the results, SIGNED by it. Those signed results are returned from ocall_verify_quote_ecdsa() here.

Then we call sgx_tvl_verify_qve_report_and_identity(), that is supposed to verify that the returned results are indeed signed by the alleged QvE (i.e. not forged or tempered with by the untrusted code).

Reply:
So `sgx_tvl_verify_qve_report_and_identity` calls the QvE directly? How do we know that the QvE is the one that performed the operation? I mean how do we know that someone didn't just hijack the function call and return a fake result?